upgrade to node 14

josephperrott opened the pull request:
Upgrades the repository to use node 14.  Additionally, removes support for node 10 as it has
reached end of life.

BREAKING CHANGE: Angular no longer maintains support for node v10



Note:  Only `core` scope commit includes a breaking change note to prevent the same note from occurring many times.

alan-agius4 commented on packages/platform-server/package.json:
```diff
@@ -34,9 +34,6 @@
     "./__ivy_ngcc__/esm2015/init/src/init.js",
     "./__ivy_ngcc__/fesm2015/init.js"
   ],
-  "engines": {
```
Why the “engines” field was removed from here?

This is useful if users are install this package in an older node.js environment, which will cause a warning or a error based on the package manager and its settings.

josephperrott replied:
Should we just set the `engines` field in `@angular/core`'s package.json since it essentially is always needed to use any of the packages? Alternatively we could set it for all packages we release.

josephperrott replied:
@jelbourn thoughts on the above?

jelbourn replied:
Seems like the "technically correct" thing is to list it on each package, no? I would probably do that unless there's some reason not too (like if it were to generate extra console noise).

The change seems to have the `engines` set on everything now, though, yes?

josephperrott replied:
Yes, this change makes the engines value appear in everyone packages `package.json`.  If someone were to be using an unsupported version of node, they would see a warning during node install for each package they rely on.

alan-agius4 reviewed (commented):
Shouldn’t the changes be split into 2 commits?

```
ci: update to use Node 14
feat(core): drop support for Node 10
```

gkalpak commented on .circleci/config.yml:
```diff
@@ -77,7 +77,7 @@ executors:
         default: medium
     docker:
-      - image: circleci/node:12.14.1@sha256:f9de24fc0017059cc42ef7d07db060008af65a98b1f0cdd1ef3339213226bf6d
+      - image: circleci/node:14.12.0
```
Previously, we deliberately specified a SHA to increase determinism of builds (see [this comment][1]).
Is this an intentional change? If so, please update the comment above (and maybe explicitly mention the reasoning in the commit message for future reference).

[1]: https://github.com/angular/angular/blob/6e3d5e0251cac5bf37fa4465e448cb22ebc0fafd/.circleci/config.yml#L69-L70

josephperrott replied:
Its not possible to look up the sha that circleci is currently using so I had to run it on CI first to get the sha to lock to.  Updated to use it.

gkalpak reviewed (changes requested):
Yesss 🎊 :tada: 🎇 
Please also update [.devcontainer/recommended-Dockerfile][1].

FWIW, I believe the changes should be split up into two commits: one that updates the version of Node.js used in the repo and another one that updates the engines in the various packages (`compiler-cli`, `localize`, `platform-server`).

[1]: https://github.com/angular/angular/blob/61bfa3d9dfc7c9daecde098aca595b731c3312a0/.devcontainer/recommended-Dockerfile#L4

devversion commented on .nvmrc:
```diff
@@ -1 +1 @@
-12.14.1
+14.16.0
```
Are we intentionally using `14.16.0` here, but above and below `14.12.0`? (e.g. for the windows CI or Bazel)

devversion replied:
Bazel seems to even support `14.16.1.` according to https://github.com/bazelbuild/rules_nodejs/blob/stable/internal/node/node_versions.bzl#L1176

josephperrott replied:
We can do `14.16.1`, I don't remember exactly how I ended up at that divergence

josephperrott committed:
build: upgrade to node 14

Upgrade local development environment for the angular repo to use node 14
dropping node 10 from supported for development within the angular repo.

josephperrott committed:
feat(core): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

BREAKING CHANGE: Angular no longer maintains support for node v10

josephperrott committed:
feat(animations): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(bazel): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(common): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(compiler-cli): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(compiler): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(elements): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(forms): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(language-service): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(localize): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(platform-browser-dynamic): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(platform-browser): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(platform-server): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(router): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(service-worker): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott committed:
feat(upgrade): update supported range of node versions

Update the supported range of node versions for Angular.  Angular now
supports node >=12.14.1 to <16.0.0, dropping support for Node v10.

josephperrott commented on packages/core/package.json:
```diff
@@ -4,6 +4,9 @@
   "author": "angular",
   "license": "MIT",
+  "engines": {
+    "node": "SUPPORTED_NODE_VERSIONS_RANGE_FOR_ENGINES_FIELD"
```
Yes, I didn't realized I had pushed up to github with this specific branch, its been removed.

devversion reviewed (approved):
LGTM. I assume we can't use `14.16.1` for Bazel yet as we aren't on the latest `rules_nodejs` version, right?

alan-agius4 reviewed (approved):
LGTM

jelbourn reviewed (approved):
LGTM

Reviewed-for: global-approvers

josephperrott commented:
Caretaker Note: This needs to be merged before feature freeze

gkalpak reviewed (approved):
LGTM ✔️

petebacondarwin commented:
> I don't quite understand why you are publishing a library that is mostly used in a browser context with the `engines` field. I could understand `platform-server`, or `compiler-cli` or `angular-cli` in general, but why the rest?

Indeed we were just discussing this. There is the slight oddity that even things like core might end up being executed inside node.js if you are doing SSR. As a team we are on the fence and slightly falling on the side of just adding the engine field to all our packages for consistency.

JounQin commented:
May I ask why Node v16 is not supported? Is there any known issue?

josephperrott commented:
@JounQin There are not any known issues, and I would even guess that it will work perfectly well.  However, Node v16 is in the `Current` release status per https://nodejs.org/en/about/releases/.  Angular only supports versions of Node which are in Active LTS or Maintenance LTS.

We should begin supporting v16 ~End of October 2021 as this is when it will begin Active LTS.

marlass commented:
@josephperrott In this case it might be worth updating that info in the docs: https://angular.io/guide/setup-local#prerequisites where `current` is mentioned in the support list.
